interpreter/cling/lib/MetaProcessor/InputValidator.cpp

```diff
@@ -15,191 +15,97 @@
+        break;
+
+      case tok::l_square: case tok::l_paren: case tok::l_brace:
+      case tok::l_comment:
+        m_ParenStack.push_back(Tok.getKind());
+        break;
+      case tok::r_square: case tok::r_paren: case tok::r_brace:
+        {
+          auto tos = m_ParenStack.empty()
+            ? tok::unknown : static_cast<tok::TokenKind>(m_ParenStack.back());
+          if (!Tok.closesBrace(tos)) {
+            Res = kMismatch;
             break;
           }
-          else if (commentTok == tok::slash) {
-            // Cancel the wait for a slash, but only if current token isn't
-            // also an asterik.
-            if (kind != tok::asterik)
-              commentTok = tok::asterik;
-          }
-        } else if (commentTok == tok::asterik) {
-          // Was wating for an asterik, but found something else.
-          // Go back to looking for a slash.
-          commentTok = tok::slash;
-        }
-
-        if (kind >= (int)tok::l_square && kind <= (int)tok::r_brace) {
-          // The closing paren kind is open paren kind + 1 (i.e odd number)
-          if (kind % 2) {
-            int prev = m_ParenStack.empty() ? -1: m_ParenStack.back();
-            // closing the right one?
-            if (prev != kind - 1) {
-              if (multilineComment)
-                continue;
-              Res = kMismatch;
-              break;
-            }
+          m_ParenStack.pop_back();
+          // '}' will also pop a template '<' if their is one
+          if (Tok.getKind() == tok::r_brace && m_ParenStack.size() == 1
+              && m_ParenStack.back() == tok::less)
             m_ParenStack.pop_back();
-
-            // Right brace will pop a template if their is one
-            if (kind == tok::r_brace && m_ParenStack.size() == 1 ) {
-              if (m_ParenStack.back() == tok::greater)
-                m_ParenStack.pop_back();
-            }
-          }
-          else
-            m_ParenStack.push_back(kind);
         }
-        else if (kind == tok::hash) {
-          MetaLexer Lex(curPos);
-          Lex.SkipWhitespace();
-          Lex.LexAnyString(Tok);
-          const llvm::StringRef PPtk = Tok.getIdent();
-          if (PPtk.startswith("endif")
-              && (PPtk.size() > 5 ? PPtk[5]=='/' || isspace(PPtk[5]) : true)) {
-            if (m_ParenStack.empty() || m_ParenStack.back() != tok::hash) {
+        break;
+
+      case tok::semicolon:
+        // Template forward declatation, i.e. 'template' '<' ... '>' ... ';'
+        if (m_ParenStack.size() == 1 && m_ParenStack.back() == tok::less)
+          m_ParenStack.pop_back();
+        break;
+
+      case tok::hash:
+        Lex.SkipWhitespace();
+        Lex.LexAnyString(Tok);
+        const llvm::StringRef PPtk = Tok.getIdent();
+        if (PPtk.startswith("if")) {
+          m_ParenStack.push_back(tok::hash);
+        } else if (PPtk.startswith("endif") &&
+                   (PPtk.size() > 5 ? PPtk[5]=='/'||isspace(PPtk[5]) : true)) {
```

Comment:
Isn't this just `PPtk.size() == 5 || PPtk[5]=='/' || isspace(PPtk[5])`, ie exact match or `/` or space? Why could there be a slash?

Reply:
Thanks for spotting, @hahnjo!  Again, from the old code...  ;-)